Добавлено домашнее задание 07: Обработчик сообщений

DmitryLapitsky opened the pull request:
Не вышло сделать тест с mock для гарантированного получения ошибки (оставлен закомиченный код, которым пользовался)
И не понял про лиснер - историю, т.к. все объекты message в проекте генерируются builder-ом, т.е. представляют собой отдельные сущности.:(

DmitryLapitsky committed:
Добавлено домашнее задание 07: Обработчик сообщений

petrelevich commented on hw07/src/test/java/ru/otus/handler/ComplexProcessorTest.java:
```diff
@@ -0,0 +1,145 @@
+//        assertThatExceptionOfType(DateTimeException.class).isThrownBy(() -> processor1.process(null));
+//        try by mock, but didnt find, how to check throw of the DateTimeException
+        while (true) {
+            if (Integer.parseInt(new SimpleDateFormat("ss").format(new Date())) % 2 == 0) {
```
это нестабильный тест.
как работать со временем посмотрите пример к паттерну мементо.

petrelevich replied:
Посмотрели пример с мементо? Видел там интерфейс для работы со временем?

DmitryLapitsky replied:
Убрал нестабильный тест, оставил только с mock. Время определяется теперь через LocalDateTime

DmitryLapitsky committed:
выкачал новую версию с обновленным заданием по historyListener.
Добавил в класс Message помимо полей еще 11-13  метод копирования объекта (лежит в Builder класса)

DmitryLapitsky committed:
должен быть удален файл класса, который уже не используется(неверная реализация ListenerHistory)

petrelevich commented on hw07/src/main/java/ru/otus/listener/HistoryListener.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public void onUpdated(Message oldMsg, Message newMsg) {
+        Message oldMsgH = new Message.Builder(oldMsg.getId()).copyBody(oldMsg).build();
+        history.put(count++, oldMsgH);
```
у Message уже есть Id, им можно пользоваться, зачем нужен count ?

DmitryLapitsky replied:
Думал вначале про реализацию - сохранять пару старое-новое значение и вести счетчик, когда какое изменение было по очереди. Убрал эту реализацию + поиск сообщения переделал под stream api

petrelevich commented on hw07/src/main/java/ru/otus/listener/HistoryListener.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public void onUpdated(Message oldMsg, Message newMsg) {
+        Message oldMsgH = new Message.Builder(oldMsg.getId()).copyBody(oldMsg).build();
```
`new Message.Builder(oldMsg.getId()).copyBody(oldMsg).build();`
лучше сделать метод clone и пользоваться им.

DmitryLapitsky replied:
Убрал из конструктора Build и перенес в класс Message метод clone()

DmitryLapitsky committed:
Реализация теста проверки ошибки на четную секунду с помощью mock

abidaJustice commented on hw07/src/main/java/ru/otus/listener/ListenerPrinter.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public void onUpdated(Message oldMsg, Message newMsg) {
+        var logString = String.format("oldMsg:%s,\nnewMsg:%s", oldMsg, newMsg);
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

DmitryLapitsky committed:
Работа над ошибками.
В HistoryListener упрощение работы с сообщениями: убран count и поиск сообщения по id через stream api
В Message добавлен метод clone()
В ProcessorEvenSecError время определяется LocalDateTime
В тестах убран нестабильный тест, оставлен только с mock.spy

petrelevich commented on hw07/src/main/java/ru/otus/listener/HistoryListener.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    public Optional<Message> findMessageById(long id) {
+        return history.stream().filter(el-> el.getId() == id).findAny();
```
если хранить не в списке, а в мапе, то можно без перебора обойтись.

petrelevich commented on hw07/src/main/java/ru/otus/model/Message.java:
```diff
@@ -0,0 +1,251 @@
+    //todo: 1. Добавить поля field11 - field13 (для field13 используйте класс ObjectForMessage)
+
+    public Message clone() {
+        ObjectForMessage copyField13 = new ObjectForMessage();
```
лучше в ObjectForMessage добавить метод clone
с класс сделать implements Cloneable

petrelevich commented on hw07/src/main/java/ru/otus/processor/ProcessorEvenSecError.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public Message process(Message message) {
+        int second = getTime();
```
посмотрите в примере к паттерну мементо, как удобно работать со временем

petrelevich commented on hw07/src/main/java/ru/otus/processor/ProcessorEvenSecError.java:
```diff
@@ -0,0 +1,22 @@
+
+public class ProcessorEvenSecError implements Processor {
+
+    public int getTime(){
```
не надо расширять интерфейс новыми методами
